[SVLS-4780] upstream Java sns and sqs refactor

joeyzhao2018 opened the pull request:
<!--- Please remember to review the [contribution guidelines](https://github.com/DataDog/datadog-lambda-js/blob/master/CONTRIBUTING.md) if you have not yet done so._  --->

### What does this PR do?
Extract trace context in SNS and SQS cases where the upstream service is instrumented by dd-trace-java
In short, 
1. For `Java => SNS => Lambda`, the env: _X_AMZN_TRACE_ID has the dd trace id but not the correct parent id. We will still use it here.
2. For `Java => SQS => Lambda` or `Java => SNS => SQS => Lambda` case, AWS propagate the trace context in the json payload attributes. 

Longer version:

![javasnssqs drawio](https://github.com/DataDog/datadog-lambda-js/assets/5253430/310a5d31-cca5-4f4b-9833-17d7e214d970)


### Testing Guidelines

- Unittests added
- Integration tests added https://github.com/DataDog/dd-trace-x-lambda-examples/blob/main/lib/wip-cases/sns-java-checks-nested-stack.ts
- Manually tested (You can see the parenting issue causing a white space between spans)
Single Message Publish
<img width="1343" alt="Screenshot 2024-05-09 at 1 24 45 PM" src="https://github.com/DataDog/datadog-lambda-js/assets/5253430/63354dda-cf04-4fa1-8845-98b1f095e308">

Batch Publish
<img width="1340" alt="Screenshot 2024-05-09 at 10 19 40 AM" src="https://github.com/DataDog/datadog-lambda-js/assets/5253430/8bd6dc8f-a257-411a-8893-8e8f59f3d7a6">

### Additional Notes

<!--- Anything else we should know when reviewing? --->

### Types of Changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [ ] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [ ] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [ ] This PR passes the integration tests (ask a Datadog member to run the tests)

duncanista commented on src/trace/context/extractors/sns-sqs.ts:
```diff
@@ -3,32 +3,46 @@ import { EventTraceExtractor } from "../extractor";
+        const messageAttribute = parsedBody?.MessageAttributes?._datadog;
+        if (messageAttribute?.Value) {
+          let headers;
+          if (messageAttribute.Type === "String") {
+            headers = JSON.parse(messageAttribute.Value);
+          } else {
+            const decodedValue = Buffer.from(messageAttribute.Value, "base64").toString("ascii");
+            headers = JSON.parse(decodedValue);
+          }
 
-      let headers;
-      if (messageAttribute.Type === "String") {
-        headers = JSON.parse(messageAttribute.Value);
-      } else {
-        const decodedValue = Buffer.from(messageAttribute.Value, "base64").toString("ascii");
-        headers = JSON.parse(decodedValue);
+          const traceContext = this.tracerWrapper.extract(headers);
+          if (traceContext) {
+            logDebug("Extracted trace context from SNS-SQS event", { traceContext, event });
+            return traceContext;
+          } else {
+            logDebug("Failed to extract trace context from SNS-SQS event", { event });
+          }
+        }
+      }
+      // Then try to extract trace context from attributes.AWSTraceHeader. (Upstream Java apps can
+      // pass down Datadog trace context in the attributes.AWSTraceHeader in SQS case)
+      if (event?.Records?.[0]?.attributes?.AWSTraceHeader !== undefined) {
+        const traceContext = XrayService.extraceDDContextFromAWSTraceHeader(event.Records[0].attributes.AWSTraceHeader);
+        if (traceContext) {
+          logDebug("Extracted trace context from SNS-SQS event attributes.AWSTraceHeader", { traceContext, event });
+          return traceContext;
+        } else {
+          logDebug("No Datadog trace context found from SNS-SQS event attributes.AWSTraceHeader", { event });
+        }
```
Is there a way to make this code more readable?

I wonder if we could create functions inside this class for specific extraction scenarios, making the `extract` look more simple. WDYT?

joeyzhao2018 replied:
I actually think it's more readable to add the logging at this level to keep it consistent as above.
The `XrayService.extraceDDContextFromAWSTraceHeader` will just try to extract dd trace context if it exists or return null. 
"Returning if found, moving on if not" is more specific at this level here.
Let me know if it makes sense?

duncanista commented on src/trace/context/extractors/sns.ts:
```diff
@@ -3,29 +3,44 @@ import { TracerWrapper } from "../../tracer-wrapper";
-      if (traceContext === null) return null;
+      // First try to extract trace context from message attributes
+      const messageAttribute = event?.Records?.[0]?.Sns?.MessageAttributes?._datadog;
+      if (messageAttribute?.Value) {
+        let headers;
+        if (messageAttribute.Type === "String") {
+          headers = JSON.parse(messageAttribute.Value);
+        } else {
+          // Try decoding base64 values
+          const decodedValue = Buffer.from(messageAttribute.Value, "base64").toString("ascii");
+          headers = JSON.parse(decodedValue);
+        }
 
-      logDebug(`Extracted trace context from SNS event`, { traceContext, event });
-      return traceContext;
+        const traceContext = this.tracerWrapper.extract(headers);
+        if (traceContext) {
+          logDebug("Extracted trace context from SNS event", { traceContext, event });
+          return traceContext;
+        } else {
+          logDebug("Failed to extract trace context from SNS event", { event });
+        }
+      }
+      // Then try to extract trace context from _X_AMZN_TRACE_ID header (Upstream Java apps can
+      // pass down Datadog trace id (parent id wrong) in the env in SNS case)
+      if (process.env[AMZN_TRACE_ID_ENV_VAR]) {
+        const traceContext = XrayService.extraceDDContextFromAWSTraceHeader(process.env[AMZN_TRACE_ID_ENV_VAR]);
+        if (traceContext) {
+          logDebug("Extracted Datadog trace context from _X_AMZN_TRACE_ID", { traceContext, event });
+          return traceContext;
+        } else {
+          logDebug("No Datadog trace context found from _X_AMZN_TRACE_ID", { event });
+        }
+      }
```
Same as the comment above! 🤔

duncanista commented on src/trace/context/extractors/sqs.ts:
```diff
@@ -10,23 +10,28 @@ export class SQSEventTraceExtractor implements EventTraceExtractor {
+        const traceContext = this.tracerWrapper.extract(JSON.parse(headers));
+        if (traceContext) {
+          logDebug("Extracted trace context from SQS event messageAttributes", { traceContext, event });
+          return traceContext;
+        } else {
+          logDebug("Failed to extract trace context from messageAttributes", { event });
+        }
       }
-      if (!parsedHeaders) return null;
-
-      const traceContext = this.tracerWrapper.extract(parsedHeaders);
-      if (traceContext === null) {
-        logDebug("Failed to extract trace context from parsed headers", { parsedHeaders, event });
-        return null;
+      // Then try to extract trace context from attributes.AWSTraceHeader. (Upstream Java apps can
+      // pass down Datadog trace context in the attributes.AWSTraceHeader in SQS case)
+      if (event?.Records?.[0]?.attributes?.AWSTraceHeader !== undefined) {
+        const traceContext = XrayService.extraceDDContextFromAWSTraceHeader(event.Records[0].attributes.AWSTraceHeader);
+        if (traceContext) {
+          logDebug("Extracted trace context from SQS event attributes AWSTraceHeader", { traceContext, event });
+          return traceContext;
+        } else {
+          logDebug("No Datadog trace context found from SQS event attributes AWSTraceHeader", { event });
+        }
```
I think we are missing to return `null` here when we cannot extract anything, should we add it?

joeyzhao2018 replied:
This is a "return if found, move on if not" block. A null will be returned at the end of the whole method if reached.

duncanista commented on src/trace/context/extractors/sqs.ts:
```diff
@@ -10,23 +10,28 @@ export class SQSEventTraceExtractor implements EventTraceExtractor {
-        return null;
+      // Then try to extract trace context from attributes.AWSTraceHeader. (Upstream Java apps can
+      // pass down Datadog trace context in the attributes.AWSTraceHeader in SQS case)
+      if (event?.Records?.[0]?.attributes?.AWSTraceHeader !== undefined) {
+        const traceContext = XrayService.extraceDDContextFromAWSTraceHeader(event.Records[0].attributes.AWSTraceHeader);
+        if (traceContext) {
+          logDebug("Extracted trace context from SQS event attributes AWSTraceHeader", { traceContext, event });
+          return traceContext;
+        } else {
+          logDebug("No Datadog trace context found from SQS event attributes AWSTraceHeader", { event });
+        }
```
```suggestion
      // log debug "trying to extract from Xray"
      if (event?.Records?.[0]?.attributes?.AWSTraceHeader !== undefined) {
          return XrayService.extraceDDContextFromAWSTraceHeader(event.Records[0].attributes.AWSTraceHeader);
        } 
```

I wonder if this can be just modified to something like

joeyzhao2018 replied:
Um.... I just think it's more critical to log afterwards on the result of `XrayService.extraceDDContextFromAWSTraceHeader`. It would tell us if it's found or not. And also directly returning is not intended here.

duncanista reviewed (commented):
Left some comments.

joeyzhao2018 commented:
> Left some comments.

@duncanista 's comments made me to take a closer look the way i added debug logs. And after some thoughts I think it's best to use just one static string as the parameter and removed extra logging of the event payload or tracecontext. This is to avoid any extra memory allocation and keep the code cleaner. Based on my on-call experience, we usually get the event payload from other logs.
